templates/cloud_init/cloud_init.client.yml

```diff
@@ -375,6 +375,63 @@ write_files:
 
       echo "DONE: end of $0"
 
+  - path: /opt/upload_gsi.sh
+    permissions: '0755'
+    content: |
+      #!/bin/bash
+
+      function error() {
+        >&2 echo "echo >< ERROR: $* See log for details."
+        exit 1
+      }
+
+      export AZURE_SUBSCRIPTION='<sub_id>'
+      export ARM_SUBSCRIPTION_ID='<sub_id>'
+      export ARM_CLIENT_ID='<arm_client_id>'
+      export ARM_CLIENT_SECRET='<arm_client_scrt>'
+      export ARM_TENANT_ID='<arm_tenant_id>'
+      export REGION='<region>'
+      export RESOURCE_GROUP_NAME='<resource_group_name>'
+      export GSI_URL='<gsi_url>'
+      export TF_VFXT_CONFIG='<terraform_file_config>'
+
+      echo "cd to file path"
+      cd /tf/src/terraform/examples/vfxt/${TF_VFXT_CONFIG}
+      [ $? -ne 0 ] && error "cd to directory: $TF_VFXT_CONFIG"
+
+      MANAGEMENT_IP=$(terraform output -json | jq -r ".management_ip.value")
+
+      echo "ssh - initial connection needed for below scp cmd"
+      ssh -o "StrictHostKeyChecking=no" -oProxyCommand="ssh -oStrictHostKeyChecking=no -W %h:%p  admin@${MANAGEMENT_IP}" root@${MANAGEMENT_IP} ls
+
+      echo "ssh to run averecmd accept terms"
+      ssh -J admin@${MANAGEMENT_IP} root@${MANAGEMENT_IP} "/support/bin/averecmd support.acceptTerms yes"
+      [ $? -ne 0 ] && error "ssh to run averecmd accept terms"
+
+      echo "ssh to run averecmd setCustomSetting"
+      ssh -J admin@${MANAGEMENT_IP} root@${MANAGEMENT_IP} "/support/bin/averecmd support.setCustomSetting 'gsiInfo.url' 'LB' '${GSI_URL}'"
+      [ $? -ne 0 ] && error "ssh to run averecmd setCustomSetting"
+
+      echo "ssh to run averecmd gsi"
+      job_id=$(ssh -J admin@${MANAGEMENT_IP} root@${MANAGEMENT_IP} "/support/bin/averecmd support.executeNormalMode cluster gsimin")
+      [ $? -ne 0 ] && error "ssh to run averecmd gsi"
+
+      response=False
+      while [ "${response}" != "True" ] ; do
+        response=$(ssh -J admin@${MANAGEMENT_IP} root@${MANAGEMENT_IP} "/support/bin/averecmd support.taskIsDone $job_id")
+        if [ ${response} ]; then
+          echo "INFO: gsi collect is still active. Re-check in 30 seconds."
+          sleep 30
+          continue
+        else
+          echo "INFO: gsi collect is  no longer active"
+          break
+        fi
+      done
```

Comment:
What is the output of `support.taskIsDone`? I would imagine it would be returning true/false (or something like done/in-progress). Does this command return nothing if a task is done?

`if [ ${response} ]` is also dangerous. What if `response` is "a b c"? Bash might interpret that as three separate tokens/strings, in which case that's a syntax error. (But this kind of check is also a potential "arbitrary code execution" point.)

Is there an example of this working? I'm not seeing any succeeded pipeline runs at this commit ID.

Reply:
support.taskIsDone returns True or False.
I had this script on a runner_vm and was testing this way and then checking the ATS-linux-1 VM to verify it was uploaded.
Will update when run completes

Reply:
Sounds good. Note that bash doesn't interpret "True" or "False" as booleans, but rather as strings.

```bash
$ if [ False ]; then echo "hello"; fi
hello
$ if [ True ]; then echo "hello"; fi
hello
$ if [ asasdas ]; then echo "hello"; fi
hello
$ if [ a b c ]; then echo "hello"; fi
bash: [: b: binary operator expected
$ if [ " " ]; then echo "hello"; fi
hello
$ if [ "" ]; then echo "hello"; fi
$ if [ ]; then echo "hello"; fi
$
```

So if $response was empty, that's treated as a boolean false. If $response has anything in it, that's boolean true.

Reply:
I want to verify I am now checking correctly:
```
  response=False
    while [ "${response}" != "True" ] ; do
      response=$(ssh -J admin@${MANAGEMENT_IP} root@${MANAGEMENT_IP} "/support/bin/averecmd support.taskIsDone $job_id")
      if [ "${response}" == "False" ]; then
        echo "INFO: gsi collect is still active. Re-check in 30 seconds."
        sleep 30
        continue
      elif [ "${response}" == "True" ]; then
        echo "INFO: gsi collect is  no longer active"
        echo "gsi_complete" > /tmp/gsi_results.done
        break
      else
        error "Response was not True or False for averecmd support.taskIsDone"
      fi
    done
      ```